### PR TITLE
TDC simulation for QIE10 (HF)

### PR DESCRIPTION
**Pull request by mseidel42**
#### PR description:

* This PR adds TDC timing simulation for the HF --> enables the use of data noise filters in MC
* Presentation in HCAL DPG: https://indico.cern.ch/event/1026044/contributions/4324030/attachments/2227968/3774461/qie10tdc_mseidel.pdf
  * Pulse shape deemed sufficient (no update)
  * Eta-dependence comes from PCaloHits in the HF shower library, will be fixed separately

#### PR validation:

MC [unit: TDC = ns/2] vs data [ns] comparison:

image

### CODE REVIEW

**Comment by abdoulline** [comment]
@mseidel42  since timePhase is changed, we may expect some  _very small_, but widespread (in many workflows) changes in HF results.

**Comment by kpedro88** [comment]
please test

**Comment by abdoulline** [comment]
@mseidel42  something went wrong with HF scale (suspiciously enough by factor 2 ) 
https://cms-cpt-software.web.cern.ch/cms-cpt-software/General/Validation/SVSuite/HCAL/calo_scan_single_pi/12_X/12_0_0_pre1_run3_PR33829_vs_12_0_0_pre1_run3_SinglePi/

**Comment by abdoulline** [comment]
@lwang046 
Long, may be you can take a look with a fresh eye at your convenience to spot where "an extra" factor 2 of  HF energy (increase) comes from after changed "dt"/"invdt" in CaloHitResponse. You've been trought it at some point with QIE11 TDC update...

File: SimCalorimetry/CaloSimAlgos/src/CaloHitResponse.cc [hunk 1]
```diff
@@ -127,13 +127,13 @@ CaloSamples CaloHitResponse::makeAnalogSignal(const PCaloHit &hit, CLHEP::HepRan
   if (storePrecise) {
     result.resetPrecise();
     int sampleBin(0);
-    // use 1ns binning for precise sample
-    for (int bin = 0; bin < result.size() * BUNCHSPACE; bin++) {
-      sampleBin = bin / BUNCHSPACE;
+    // use 0.5ns binning for precise sample
+    for (int bin = 0; bin < result.size() * BUNCHSPACE * invdt; bin++) {
+      sampleBin = bin / (BUNCHSPACE * invdt);
       double pulseBit = (*shape)(binTime)*signal;
```
Comment (mseidel42) [inline SimCalorimetry/CaloSimAlgos/src/CaloHitResponse.cc:133]: Hi @abdoulline, maybe this line needs a factor `dt`?

Comment (abdoulline) [reply SimCalorimetry/CaloSimAlgos/src/CaloHitResponse.cc:133]: I recall this change by Long  (in https://github.com/cms-sw/cmssw/pull/29310), which we've been discussing with Kevin back then
https://github.com/cms-sw/cmssw/pull/29310/files#diff-ebeb6f4661dfc9b0c622167d87df27846dc8ef5134548f04e4ed411555ec44dbR257

Comment (abdoulline) [reply SimCalorimetry/CaloSimAlgos/src/CaloHitResponse.cc:133]: @mseidel42   I think so... Line #133  needs "dt" factor, as the sampling is twice more "often" now, so each bin content should be now twice smaller.  Let me try to check/test it explicitly with CaloScan.

Comment (mseidel42) [reply SimCalorimetry/CaloSimAlgos/src/CaloHitResponse.cc:133]: Ok, thanks for checking! I am also rerunning my local test at the moment.

**Comment by abdoulline** [comment]
@mseidel42   OK now, 1-2% scale diff. 
https://cms-cpt-software.web.cern.ch/cms-cpt-software/General/Validation/SVSuite/HCAL/calo_scan_single_pi/12_X/12_0_0_pre1_run3_PR33829_mod_vs_12_0_0_pre1_run3_SinglePi/

**Comment by mseidel42** [comment]
Thank you, that's more reasonable :-)

Will add this then:
```diff
-      double pulseBit = (*shape)(binTime)*signal;
+      double pulseBit = (*shape)(binTime)*signal*dt;
```

**Comment by civanch** [comment]
please test

**Comment by kpedro88** [comment]
I was concerned about:
> You potentially added 19030 lines to the logs

These appear to come from the premix workflows 250202.181 and 34834.999, where there are many instances of this message in step3:
```
%MSG
%MSG-e CaloHitResponse:  PreMixingModule:mixData 26-May-2021 15:10:49 CEST  Run: 1 Event: 1
Mismatched calo signals 
%MSG
```

This may just imply that the premixed pileup samples need to be regenerated after this PR is merged. But just to be sure, I think it would be a good idea to compare:
1. Baseline results from premixed workflow
2. PR results from premixed workflow w/ old premixed pileup sample
3. PR results from premixed workflow w/ new (regenerated) premixed pileup sample

**Comment by abdoulline** [comment]
@kpedro88   Kevin,  I believe it now becomes quite obvious (while I must admit I didn't think of it prior to this PR) that "old" premixed PU is clearly incompatible with "new" (more fine-grain) HF preciseSamples ?

**Comment by kpedro88** [comment]
That makes sense; it's just been a long time since I looked at the HF premixing implementation, so I would personally err on the side of caution and check the validity of the results.

**Comment by abdoulline** [comment]
@kpedro88 
if I look (e.g.) at  250202.181 [*]  
I can see that premixed file ("step2" from SingleNuE10_cfi_GEN_SIM_DIGI_PU.py ) is the **new/regenerated** one from Minbias GENSIM (from disk) input. So it should supposedly be HF-compatible with the signal ("step1" from TTbar_13TeV_TuneCUETP8M1_cfi_GEN_SIM.py), unless I'm missing something.
Or there could be something else affected...
-------
[*]
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-349617/15318/runTheMatrix-results/250202.181_TTbar_13UP18+TTbar_13UP18+PREMIXUP18_PU25+DIGIPRMXLOCALUP18_PU25+RECOPRMXUP18_PU25+HARVESTUP18_PU25/

**Comment by abdoulline** [comment]
I've looked at  250202.181 (10 ev)  HF results  for 
(i) regular 12_0_0_pre1 
(ii) idem + this PR
And there is just a very small (expected) difference. So HF "per se" looks OK, but something else may not be... 

(i)  regular 
RecHits_HF_spectrum_std

(ii) + PR
RecHits_HF_spectrum_upd

**Comment by mseidel42** [comment]
Hi Kevin, hi Salavat,

Ok, I can do some debugging using 250202.181 next week.

Best,
Markus

**Comment by abdoulline** [comment]
https://cmssdt.cern.ch/lxr/source/CalibFormats/CaloObjects/src/CaloSamples.cc#0053
possibly to add: 

(1) HcalDetId  - header to code and DataFormats/HcalDetId to BuildFile  
(2)
DetId id = other.id();
if (id.det == DetId::Detector::Hcal)  {  //  HCAL :  https://cmssdt.cern.ch/lxr/source/DataFormats/DetId/interface/DetId.h#0028
    HcalDetId hid = HcalDetId(id);
    int sub = hid.subdet();   // .ieta(), .iphi(), .depth()    
    cout << "HCAL subdet " << sub  <...> 
}
else  // not HCAL 
    cout << Not HCAL  <...>

**Comment by mseidel42** [comment]
Hi Salavat, hi Kevin,

I added printouts and found that some of the mismatches even come from HB:
```
HcalDetId 1126187065, subdet 1, 8 samples, 400 preciseSamples, 50 preciseStep                                                                                      
...
HcalDetId 1126187065, subdet 1, 8 samples                                                                                                                          
```

Plus mismatches from HF, of course.
But there are also many other CaloSamples from both HB and HF matching correctly, leaving me a bit confused.

**Comment by abdoulline** [comment]
@mseidel42   for wf 250202.181 (2018)  both HF _and_ HB (yet HPD QIE8)  processed similarly, so both affected...
I must admit I'm not expert in the premixing domain, so hope Kevin remembers some relevant particularities there.
To me it looks like a change of preciseSamples bin size was incomplete, well - as Kevin assumed above (even thought I don't immediately see what he means about this QIE10 fragment). So seems that under some circumstances it's not effective.
You may need to add some more printouts in the premixing processing chain (including the code pointed out by Kevin) to possibly spot an omission...

**Comment by mseidel42** [comment]
Hi Salavat, thank you for pointing out that HB is still QIE8 in that workflow, that eased my mind :-)

I found that in addition to the lines that Kevin pointed me to, CaloSamples were sometimes reset later in `HcalCoderDb::adc2fC_`.

Things seem to work now but I think there are 2 big questions:
* Why does this problem not appear for QIE11? If the QIE11 solution is more elegant than my fix, can we apply it to QIE8/10 as well?
* Is there any preciseData stored for pileup at all?

**Comment by abdoulline** [comment]
Markus,  afaik QIE11 simulation is somewhat separately set/processed, Long @lwang046 have intervened there in  SimCalorimetry/HcalSimAlgos/src/HcalSiPMHitResponse.cc 
https://github.com/cms-sw/cmssw/pull/29310/files

I hope Kevin (still) remebers it by heart and could answer your questions.

**Comment by abdoulline** [comment]
Concerning PU - looking at the available HE TDC plot in wf  250202.181 (in question)  one can see that TDC is simulated in all 8TS's  (Y axis: 25n * i_TS + TDC*0.5).
Means PU is involved:
https://cmsweb.cern.ch/dqm/dev/plotfairy/overlay?session=1iMGKj;v=1621959688069731899;obj=archive%2F1%2FRelVal_wf250202_181_pr%2FCMSSW_12_0_X-PRcmssw_33829-15293%2FDQMIO%2FHcalDigisV%2FHcalDigiTask%2FHcalDigiTask_TDCtime_vs_ADC_HE;obj=archive%2F1%2FRelVal_wf250202_181_base%2FCMSSW_12_0_X-PRcmssw_33829-15293%2FDQMIO%2FHcalDigisV%2FHcalDigiTask%2FHcalDigiTask_TDCtime_vs_ADC_HE;reflabel=1;norm=False;w=1120;h=580

**Comment by abdoulline** [comment]
@kpedro88 
Kevin, we need your expert's insight - 
how premixing works "out of the box" for QIE11, e.g. here (without preciseSize for CaloSamples) ? 
https://cmssdt.cern.ch/lxr/source/SimCalorimetry/HcalSimAlgos/src/HcalSignalGenerator.cc#0090

@mseidel42  
Markus,  I'm afraid you may **not** simply comment  an initialization of CaloSamples in HcalCoderDB: 
https://github.com/cms-sw/cmssw/pull/33829/commits/a8925e3eee52d19e8eeb9f8336e7231cd3ec2f0c#diff-79f226a3ba580d0637feb6987ee0a38ba16c258fbe4924ead030e7dfa9b9dcf6L14

it just provokes crashes in many places (in many wf's), where new CaloSamples is created in HcalCoderDb::adc2fC_ 
(and not used already earlier defined ones), e.g.   
https://cmssdt.cern.ch/lxr/source/RecoLocalCalo/HcalRecProducers/src/HBHEPhase1Reconstructor.cc#0504

**Comment by abdoulline** [comment]
@mseidel42  
although not directly related to the issues considered here,  here is an addition of 2 HF TDC plots to MC Digis validation: 
https://github.com/abdoulline/cmssw/tree/Validation_HF_TDC_plots  -> 
https://github.com/abdoulline/cmssw/commit/82946ca63fedd995df29b4f8164b3c5e558cc251

(git cms-merge-topic abdoulline:Validation_HF_TDC_plots)
May I ask you to add it eventually to this PR, when major issues will be sorted out?

**Comment by abdoulline** [comment]
@mseidel42 
the following simplest change seems to fix the issue, as it works in both cases (whether CaloSamples input is already fully set or not yet). At least for tested wf 250202.181 (in question)...

clf = CaloSamples(df.id(), df.size(), clf.preciseSize());

in both lines :

https://github.com/cms-sw/cmssw/pull/33829/files#diff-79f226a3ba580d0637feb6987ee0a38ba16c258fbe4924ead030e7dfa9b9dcf6R14

https://github.com/cms-sw/cmssw/pull/33829/files#diff-79f226a3ba580d0637feb6987ee0a38ba16c258fbe4924ead030e7dfa9b9dcf6R23

**Comment by abdoulline** [comment]
Using  PR + aforementioned fix + HF TDC plots (branch)  in  wf 250202.181 

HF_TDC_vs_ADC

HF_TDC

I'm a bit confused, as there seem to be no real HF TDC (means all 0) in the plots, unless I'm missing something...

**Comment by abdoulline** [comment]
Well,  I was probably too fast... 
If CaloSamples with **filled** preciseData comes/enters  HcalCoderDb::adc2fC_()  it will be re-created _without_ preciseData... 
So, there should rather be something like this (instead) :

  if ( clf.preciseSize() == 0 ) clf = CaloSamples(df.id(), df.size());   // only in case clf didn't contain preciseData

**Comment by kpedro88** [comment]
@abdoulline:
> @kpedro88
Kevin, we need your expert's insight -
how premixing works "out of the box" for QIE11, e.g. here (without preciseSize for CaloSamples) ?
https://cmssdt.cern.ch/lxr/source/SimCalorimetry/HcalSimAlgos/src/HcalSignalGenerator.cc#0090

For QIE11, preciseSamples are only used in premix stage 1 (constructing the pre-mixed pileup "digis"). The "result" from premix stage 2 is just a normal CaloSamples.

I think you're narrowing down the location of the problem, but a general suggestion: it could be useful to change this line to an exception, in order to trace back whenever any mismatch occurs. https://cmssdt.cern.ch/lxr/source/CalibFormats/CaloObjects/src/CaloSamples.cc#0053

**Comment by abdoulline** [comment]
@kpedro88  thanks!
Actually Markus has already changed ("reinforced") this mismatch warning to SIGSEGV _specifically_ for HCAL channels:
 https://github.com/cms-sw/cmssw/pull/33829/files#diff-de6203b23987e9a0a493ae473a3dd6b6d17203c79327ce3ef36b34daf5f728b3R59

I didn't see this issue (neither LogError for any mismatched channel prior to it) in my tests today...
 
The problem is that in _unpacked_ Digis (step4) HF TDCs are all zeros in wf 250202.181 
(even after I've updated HcalCoderDb::adc2fC_()  as indicated in my previous message):

HF Digi(HF 39,11,4) 3 samples
  ADC=3 TDC(LE)=0 TDC(TE)=0 CAPID=0
  ADC=31 TDC(LE)=0 TDC(TE)=0 CAPID=1 SOI
  ADC=5 TDC(LE)=0 TDC(TE)=0 CAPID=2

While a simple particle gun yields reasonable results: 

HF Digi(HF 33,53,1) 3 samples
  ADC=5 TDC(LE)=63 TDC(TE)=0 CAPID=1
  ADC=23 TDC(LE)=13 TDC(TE)=0 CAPID=2 SOI
  ADC=4 TDC(LE)=63 TDC(TE)=0 CAPID=3

@mseidel42  
Markus, I'm afraid now your turn... You may need to add more debugging printouts specificaly for preciseData, 
e.g. 
(i) here  (where samples are added)  
https://github.com/cms-sw/cmssw/pull/33829/files#diff-de6203b23987e9a0a493ae473a3dd6b6d17203c79327ce3ef36b34daf5f728b3R59
and 
(ii) at the entrance of HcalTDC

**Comment by mseidel42** [comment]
Thank you @abdoulline, I missed the HcalCoderDb crashes because I was only testing with `step3_DIGI_DATAMIX_L1_DIGI2RAW_HLT.py`. I will also include your validation updates!
But I have the feeling that patching the QIE10 CaloSamples at various places is not the most elegant way, it would be better to have it in a way similar to QIE11.

@kpedro88 Do you have some slides or PR discussion how the preciseData was handled for QIE11?

**Comment by kpedro88** [comment]
@mseidel42 the usage of preciseSamples in QIE11 is very different from what you propose here for QIE10, as I remarked to @abdoulline above. #17920 has more details.

@abdoulline I'm quite puzzled that the TDC information is kept in a particle gun workflow but lost in 250202.181... usually I would look for a problem in the packer or unpacker if information was missing in the digis, but that should not depend on the workflow (unless the workflows use different Eras).

**Comment by abdoulline** [comment]
@kpedro88   
That's a good point about different eras!  
Mea culpa, I've compared apples (pion gun 2021) to oranges (wf 250202.181)...  
Now that I ran pion gun specifically for 2018,  it yileds HF all TDC=0.   
Even though I don't remeber any changes for HF in 2018->2021...
Let me look at other wfs.

**Comment by abdoulline** [comment]
@mseidel42  
I've activated TF TDC for 2018  [1]  (with HcalCoderDb::adc2fC_() updated [2])  
and HF TDCs appear after unpacker, but the patterns don't look OK to me [**].
I'm afraid premixing deseves some low-level investigation, as I've suggested previously...

-------------------------------------------------------------------------------------------------
[1]
in  https://github.com/cms-sw/cmssw/blob/a8925e3eee52d19e8eeb9f8336e7231cd3ec2f0c/SimCalorimetry/HcalSimProducers/python/hcalSimParameters_cfi.py#L146

from Configuration.Eras.Modifier_run2_HCAL_2018_cff import run2_HCAL_2018
run2_HCAL_2018.toModify( hcalSimParameters,
    hf1 = dict( threshold_currentTDC = cms.double(3.) ),
    hf2 = dict( threshold_currentTDC = cms.double(3.) )
)


[2]
if ( clf.preciseSize() == 0 ) clf = CaloSamples(df.id(), df.size()); // QIE8 - only in case clf didn't contain preciseData
if ( clf.preciseSize() == 0 ) clf = CaloSamples(df.id(), df.samples()); // QIE10 - idem

[3]

HF Digi(HF -40,51,1) 3 samples
  ADC=55 TDC(LE)=63 TDC(TE)=0 CAPID=0
  ADC=45 TDC(LE)=18 TDC(TE)=0 CAPID=1 SOI
  ADC=40 TDC(LE)=63 TDC(TE)=0 CAPID=2

HF Digi(HF 41,43,3) 3 samples
  ADC=47 TDC(LE)=63 TDC(TE)=0 CAPID=0
  ADC=81 TDC(LE)=63 TDC(TE)=0 CAPID=1 SOI
  ADC=36 TDC(LE)=63 TDC(TE)=0 CAPID=2

HF Digi(HF 41,47,2) 3 samples
  ADC=23 TDC(LE)=63 TDC(TE)=0 CAPID=0
  ADC=72 TDC(LE)=63 TDC(TE)=0 CAPID=1 SOI
  ADC=23 TDC(LE)=63 TDC(TE)=0 CAPID=2

HF Digi(HF -41,15,1) 3 samples
  ADC=54 TDC(LE)=63 TDC(TE)=0 CAPID=0
  ADC=41 TDC(LE)=18 TDC(TE)=0 CAPID=1 SOI
  ADC=49 TDC(LE)=63 TDC(TE)=0 CAPID=2

**Comment by mseidel42** [comment]
Hi Salavat,

From Kevin's PR and presentation (thanks for the pointer!) I understand now that we are storing "lossless" 0.5 ns samples in the premix QIE11DataFrames.
Let me see if this could be done for QIE10 as well :-)
(The unfavorable alternative would be to ignore PU contributions to the precise samples and thus to the QIE10 TDC)

Best,
Markus

**Comment by abdoulline** [comment]
@mseidel42    
Hm...m... I thought (mistakenly?  - I didn't go into details though) that for QIE10 (initially for QIE8)  there existed (albeit not used since it's implementation ~10 years ago)  a similar preciseData (whatever is its bin size) summing in PU premixing -> mixing with singnal...  Even if it's not done the same way as for SiPMs QIE11. That's what/why I was asking you to check for HF QIE10 - how well this summation is done...

Anyway, it would certainly be good to have for QIE10 similar results (even if not the very same way) as for QIE11. 
But if my memory serves me well,  for HF MC filter exercises, one (Alexander T.)  just need _some_ non-0 TDC values for SOI. And none plans to rely on HF TDC simulation in general for any other purpose?   I mean the goal is rather modest, I suppose.

**Comment by mseidel42** [comment]
Hi Salavat, it seems not, this is what I get from the premix file:
qie11digis.samples(): 400
qie10digis.samples(): 3

One of the motivations for a proper TDC simulation would be that it would give us false positive filter results in the MC. If we decide to neglect PU for that we can follow the trajectory in this PR (adding empty preciseData where needed). Still, following the QIE11 implementation would give more correct results.

**Comment by abdoulline** [comment]
OK,  Markus, I see.
 So, you'll try to add preciseData filling & storing to PU premixing  (step2 in the wf in question)
https://cmssdt.cern.ch/lxr/search?%21&_filestring=&_string=PreMix1&_casesensitive=1

So as to have signal mixing (step3 in our case) with    qie10digis.samples(): 150 
https://cmssdt.cern.ch/lxr/search?%21&_filestring=&_string=PreMix2&_casesensitive=1

**Comment by tvami** [comment]
Hi @mseidel42 originally I thought you wanted this to be included in the Run-3 simulations, so essentially have a backport to 12_0_X . However, 12_0_X was cut today and we'll have 12_0_0 out by the end of the week, and simulation production is to be launched in the next weeks. May I ask what are the HCAL plans for this PR, having in mind that it cannot be included in 12_0_X anymore?

**Comment by tvami** [comment]
Hi @mseidel42 last open 12_1_X pre-release is closing in about 2 weeks from now (2021/10/5). Do you think it's realistic to aim for that?

**Comment by tvami** [comment]
@cmsbuild , please test

 * just to see what the current state is like

**Comment by mseidel42** [comment]
Hi Tamas, I have ported the current state to 12_1.
Will try to implement the fine-grained premix DIGIs today.

**Comment by mseidel42** [comment]
Close in favor of 12_1 PR: https://github.com/cms-sw/cmssw/pull/35308

**Comment by tvami** [comment]
-alca 

 * PR was closed

**Pull request closed without merging**